sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -151,7 +159,7 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]]
     // SPARK-32753: it only makes sense to copy tags to a new node
     // but it's too expensive to detect other cases likes node removal
     // so we make a compromise here to copy tags to node with no tags
-    if (tags.isEmpty) {
+    if ((_tags eq null) || _tags.isEmpty) {
       tags ++= other.tags
     }
```

Comment:
nit: maybe add an `(...) && other.tags.nonEmpty` check too?

Reply:
Good call; i abstracted this emptiness check out.